Add alias for block content to allow injections.

kristijanhusak opened the pull request:
This will allow highlighting `BLOCK_SRC` blocks with treesitter injections.

kristijanhusak committed:
Add alias for block content to allow injections.

emiasims reviewed (commented):
`block`, `dynamic_block`, and `latex_env` all use `seq(repeat($._nl), repeat(seq(repeat1($._text), repeat1($._nl)))),`, so I think we can just refactor that out to `contents`. Or `textbody`, or some other name. Perhaps filed under the Text section? Would you mind expanding the pr for those changes?

Additionally, I'm guessing you have some auto-remove trailing whitespace option, can you remove the changes for lines 1188 and 1196?

kristijanhusak commented:
I tried doing it (diff below), but I'm getting this error:

```
The rule `contents` matches the empty string.

Tree-sitter does not support syntactic rules that match the empty string
unless they are used only as the grammar's start rule.
```
I'm not sure how to tackle this. Can you assist?

Here's the diff:
```diff
diff --git a/grammar.js b/grammar.js
index 2e851b1..2d2262b 100644
--- a/grammar.js
+++ b/grammar.js
@@ -391,8 +391,7 @@ org_grammar = {
     block: $ => seq(
       optional($._directive_list),
       $._block_begin,
-      repeat($._nl),
-      alias(repeat(seq(repeat1($._text), repeat1($._nl))), $.content),
+      $.contents,
       $._block_end,
     ),
 
@@ -415,8 +414,7 @@ org_grammar = {
     dynamic_block: $ => seq(
       optional($._directive_list),
       $._dynamic_begin,
-      repeat($._nl),
-      repeat(seq(repeat1($._text), repeat1($._nl))),
+      $.contents,
       $._dynamic_end,
     ),
 
@@ -488,8 +486,7 @@ org_grammar = {
     latex_env: $ => seq(
       optional($._directive_list),
       $._env_begin,
-      repeat($._nl),
-      repeat(seq(repeat1($._text), repeat1($._nl))),
+      $.contents,
       $._env_end,
     ),
 
@@ -527,6 +524,10 @@ org_grammar = {
       seq($._text, token.immediate('_'), token.immediate(/[^\p{Z}]/))
     )),
 
+    contents: $ => seq(
+      repeat($._nl),
+      repeat(seq(repeat1($._text), repeat1($._nl))),
+    ),
   }
 };
```

milisims commented:
Pretty common 'gotcha' for writing these parsers, as the message says you can't have a rule that matches an empty string, and here you have a sequence of two rules that can match nothing (`repeat`), so I think the solution that makes the most sense is in `contents` require the second repeat (`repeat1(seq(...))`) and make the references to contents optional: `optional($.contents)`

kristijanhusak committed:
Extract block content into separate node and use it for blocks and latex nodes.

kristijanhusak commented:
That fixed it, thanks!
I extracted and fixed few tests (also reverted the whitespace, sorry about that), but `2. LatexEnv.3  -:` now considers empty block an error. What do you want me to do about it?

```
  2. LatexEnv.3  -:
    (document (ERROR)) (body (latex_env)))
```

emiasims commented on grammar.js:
```diff
@@ -527,6 +524,10 @@ org_grammar = {
       seq($._text, token.immediate('_'), token.immediate(/[^\p{Z}]/))
     )),
 
+    contents: $ => seq(
+      repeat($._nl),
+      repeat1(seq(repeat1($._text), repeat1($._nl))),
+    ),
```
If you change this bit to:
``` javascript
    contents: $ => seq(
      repeat($._text),
      sep1(repeat1($._nl), repeat1($._text)),
    ),
```
then we should be good to go.
I think this is inconsistent with `$.body`, because that one only starts at the actual text, but... that's fine.

kristijanhusak replied:
That fixed it.
